pkg/controller/master/backup/backup.go

```diff
@@ -280,65 +332,275 @@ func (h *Handler) getBackupPVC(namespace, name string) (*corev1.PersistentVolume
 	return pvc, nil
 }
 
-func (h *Handler) updateStatus(vmBackup *harvesterv1.VirtualMachineBackup, source *kv1.VirtualMachine) error {
-	var vmBackupCpy = vmBackup.DeepCopy()
-	if vmBackupCpy.Status == nil {
-		vmBackupCpy.Status = &harvesterv1.VirtualMachineBackupStatus{
-			ReadyToUse: pointer.BoolPtr(false),
+// reconcileVolumeSnapshots create volume snapshot if not exist.
+// For vm backup from a existent VM, we create volume snapshot from pvc.
+// For vm backup from syncing vm backup metadata, we create volume snapshot from volume snapshot content.
+func (h *Handler) reconcileVolumeSnapshots(vmBackup *harvesterv1.VirtualMachineBackup) error {
+	vmBackupCpy := vmBackup.DeepCopy()
+	for i, volumeBackup := range vmBackupCpy.Status.VolumeBackups {
+		if volumeBackup.Name == nil {
+			continue
+		}
+
+		snapshotName := *volumeBackup.Name
+		volumeSnapshot, err := h.getVolumeSnapshot(vmBackupCpy.Namespace, snapshotName)
+		if err != nil {
+			return err
+		}
+
+		if volumeSnapshot == nil {
+			volumeSnapshot, err = h.createVolumeSnapshot(vmBackupCpy, volumeBackup)
+			if err != nil {
+				logrus.Errorf("create volume snapshot error: %v", err)
+				return err
+			}
+		}
+
+		if volumeSnapshot.Status != nil {
+			vmBackupCpy.Status.VolumeBackups[i].ReadyToUse = volumeSnapshot.Status.ReadyToUse
+			vmBackupCpy.Status.VolumeBackups[i].CreationTime = volumeSnapshot.Status.CreationTime
+			vmBackupCpy.Status.VolumeBackups[i].Error = translateError(volumeSnapshot.Status.Error)
 		}
-	}
 
-	if source != nil {
-		vmBackupCpy.Status.SourceUID = &source.UID
 	}
 
-	if isBackupProgressing(vmBackupCpy) {
-		source, err := h.getBackupSource(vmBackupCpy)
-		if err != nil {
+	if !reflect.DeepEqual(vmBackup.Status, vmBackupCpy.Status) {
+		if _, err := h.vmBackups.Update(vmBackupCpy); err != nil {
 			return err
 		}
+	}
 
-		if source != nil {
-			updateBackupCondition(vmBackupCpy, newProgressingCondition(corev1.ConditionTrue, "Operation in progress"))
-		} else {
-			updateBackupCondition(vmBackupCpy, newProgressingCondition(corev1.ConditionFalse, "Source does not exist"))
+	return nil
+}
+
+func (h *Handler) getVolumeSnapshot(namespace, name string) (*snapshotv1.VolumeSnapshot, error) {
+	snapshot, err := h.snapshotCache.Get(namespace, name)
+	if apierrors.IsNotFound(err) {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	return snapshot, nil
+}
+
+func (h *Handler) createVolumeSnapshot(vmBackup *harvesterv1.VirtualMachineBackup, volumeBackup harvesterv1.VolumeBackup) (*snapshotv1.VolumeSnapshot, error) {
```

Comment:
u will also need to check if the `volumeSnapshot` exist on restore otherwise the PVC may failed to create with the following case:
1. config the backup target to another one
2. delete vmbackups and its volumesnaphost is not removed due to `bound-protection` finalizer
3. re-config backup target to the original one
4. the VMbackup CR is restored but the volumesnapshot is missing(will be delete after a period of seconds)

Reply:
I check whether `volumeSnapshot` has `DeletionTimestamp`. If yes, we will requeue `vmBackup` again.